lib/Sema/CSGen.cpp

```diff
@@ -356,7 +389,183 @@ namespace {
     /// Ignore types.
     bool walkToTypeLocPre(TypeLoc &TL) override { return false; }
   };
-  
+
+  SmallVector<bool, 2> literalArgsForExpr(ApplyExpr *opExpr) {
+    SmallVector<bool, 2> literalArgs;
+    if (auto tuple = dyn_cast<TupleExpr>(opExpr->getArg())) {
+      for (auto arg : tuple->getElements())
+        literalArgs.push_back(arg->getKind() >= ExprKind::First_LiteralExpr &&
+                              arg->getKind() <= ExprKind::Last_LiteralExpr);
+    } else {
+      literalArgs.push_back(
+          opExpr->getArg()->getKind() >= ExprKind::First_LiteralExpr &&
+          opExpr->getArg()->getKind() <= ExprKind::Last_LiteralExpr);
+    }
+    return literalArgs;
+  }
+
+  bool isTypeInCollectedTypes(ConstraintSystem &CS, LinkedTypeInfo &lti,
+                              TypeBase *type, bool literalArg) {
+    if (lti.collectedTypes.count(type))
+      return true;
+
+    if (auto generic = type->getAs<GenericTypeParamType>()) {
+      // If the generic param is an associated type or we have literals
+      // that may be lots of types, some of which may or may not
+      // conform, then this is too complicated, and assume it's
+      // possible.
+      if (literalArg || generic->getDepth() || generic->getIndex())
+        return true;
+
+      if (auto proto =
+              generic->getDecl()->getDeclContext()->getSelfProtocolDecl()) {
+        for (auto ty : lti.collectedTypes)
+          if (TypeChecker::conformsToProtocol(
+                  ty, proto, CS.DC, ConformanceCheckFlags::InExpression)
+                  .hasValue())
+            return true;
+      }
+      return false;
+    }
+
+    if (literalArg) {
+      for (auto proto : lti.literalProtocols(CS.TC.Context))
+        if (TypeChecker::conformsToProtocol(type, proto, CS.DC,
+                                            ConformanceCheckFlags::InExpression)
+                .hasValue())
+          return true;
+    }
+
+    for (auto convertTy : lti.collectedTypes) {
+      if (CS.TC.isConvertibleTo(convertTy, type, CS.DC))
```

Comment:
This `isConvertiableTo` (which creates new constraint system every time) and `conformsToProtocol` are pretty heavy checks which, in my opinion, we are better off only performing while solving for each individual overload. 

There is a long standing improvement opportunity where `determineBestBindings` could actually check whether found type conforms to all of the protocols associated with its type variable via `{Literal}ConformsTo` constraint and avoid that binding.